lib/utils.bash

```diff
@@ -7,6 +7,45 @@ GREP_COLORS=
 
 ASDF_DIR=${ASDF_DIR:-''}
 ASDF_DATA_DIR=${ASDF_DATA_DIR:-''}
+BASH_MAJOR_VERSION=${BASH_VERSION%%.*}
+
+fast_basename() {
+  # if the path ends in a slash, remove it, then use parameter substitution
+  # if it doesn't, use parameter substitution on the path as-is
+  dirpath="$1"
+  [[ "$dirpath" == */ ]] && dirpath="${dirpath%/}" && dirpath="${dirpath##*/}" || dirpath="${dirpath##*/}"
+  printf "$dirpath"
+}
+
+fast_tr() {
+  inputArr=("$@")
+  inputStr="${inputArr[0]}"
+  local outputStr
+  if [ ! ${#inputArr[@]} -eq 3 ] && [ ! ${#inputArr[@]} -eq 5 ]; then
+    printf "%s\n" "ERROR: fast_tr expects an array of 3 or 5 elements, got ${#inputArr[@]}"
+    return 1
+  fi
+  # if bash >= v4, use parameter modification to upper/lower the string
+  if [ ${BASH_MAJOR_VERSION} -gt 3 ]; then
+    case ${inputArr[1]} in
+      "[:lower:]")
+        outputStr="${inputStr@U}"
+        ;;
+      "[:upper:]")
+        outputStr="${inputStr@L}"
```

Comment:
This breaks since `@U` and `@L` were added in bash 5.0 while `,,` and `^^` were added in bash 4.0. But the check is still good because we support 3.2 at a minimum.

Reply:
Ah, good catch. I originally had the `^^` syntax but then found out about `@U` and used it since it seemed more intuitive. I didn't check the version requirements for it, though.